### DISCUSSION TITLE
Questions about tncattach

### DISCUSSION THREAD

**Discussion by SpudGunMan** [markqvist/Reticulum#159] [Help & Questions] [answered]
if so .. my Q is 
most things use `AWG` to connect xastir for example supports native but not kiss2TCP, how do you shim your app into the older TNC apps? I cant see to get things to function

if this s wrong spot please just close :)

**Comment by markqvist** [comment]
Fine to ask here, `tncattach` is such a small program it doesn't have its own forum section.

Can you explain more about what you are trying to achieve exactly? The `tnattach` program lets you attach TNC devices as generic Ethernet/IP interfaces. It is not for using with "legacy" AX.25 apps, but for using KISS TNCs as Ethernet adapters / IP bridges.

**Comment by SpudGunMan** [comment]
Yup understood, I was wondering if there is any legacy functionality at all (awg emulation) . I want to have a rpi on serial tnc useable by apps commonly encountered  (xastir, winlink, terminal) or if this is only a kissattach replacement?

was wondering if you had any ideas on it. 

Direwolf emulation is spot on for legacy apps, however passing off the packet to a chip via Bt is also cool not sure any facility for serial tnc

**Reply by markqvist** [reply]
Ah, like that. No, `tncattach` will not replace the kernel AX.25 functionality in any way. It is only meant as a way to attach KISS TNCs as native (and user-accessible) network interfaces.

**Comment by SpudGunMan** [comment]
Isn’t the Functionality of the ax stack and kissattach to bring kiss and radio to tcp for legacy protocol, I’m confused now, your app lights up tcp for .. kiss but what’s the use case? If not for any legacy packet apps 

Why is this program exist just for theoretical p2p links? Have you set this up with a radio and tnc yourself? What p2p links is this useful for if not capable in aprs or 1200b vhf?

**Reply by markqvist** [reply]
Legacy packet apps don't use TCP in any way, they use AX.25. This is the case for all apps like Xastir, Winlink, etc.

Practically all apps allow you to connect directly to a TNC using KISS framing over a serial port or similar. Some apps allow you to connect to the TNC itself over TCP, ie., the TNC uses a TCP connection to expose it's KISS port, instead of a USB/serial connection.

Most legacy packet radio apps can also use AX.25 support built into the Linux kernel. In that case, the user-space programs talk to the kernel interface, usually called something like `ax0` or `ax1`, and the kernel coordinates with the actual hardware TNC using KISS framing over either USB, serial or TCP. 

Windows did not have actual kernel support for AX.25, so a clever op, SV2AGW, made the `AGW` program as a hack for Windows to allow multiple programs to share the same physical/virtual TNC. Linux never needed this, since there was direct support for AX.25 and device/app multiplexing in the kernel.

But now most distros are dropping the AX.25 layers from the kernel because the implementations are old, quite messy at this point, and more or less unmaintained.

This means, that while most legacy packet radio apps will still work when *directly* talking to a TNC, it is no longer possible to set up a multiplexed environment with many apps talking to the same TNC, like with the old `ax25-utils` included in Linux, until recently.

Reticulum of course let's you do this, and in a way much more flexible than the AX.25 stack ever did.

The `tncattach` program was never meant to remedy this situation. In fact, it was more meant to help transition *away* from AX.25.

While I have great respect for the pioneering work of AX.25, and all the utility and usefulness it has provided, I would also suggest moving away from it entirely by now.

In AX.25, there is no *actual* routing, it is very inefficient with bandwidth, it is completely unreliable, since no packets can be verified or authenticated, it is completely open to modification of traffic, eavesdropping and tampering, you must have complete and full trust in *all* nodes on the network, and many more problems that are simply a deal-breaker for any real-world use, even for hobbyists.

I think a much better solution is to use Reticulum, and help transition applications and use-cases to a much more modern, flexible and efficient packet-oriented stack instead.

The `tncattach` program allows you to run full Ethernet over packet radio, or simpler IP-only P2P links if that is more suitable to the situation. That is the intention, and what it delivers.

I am not trying to salvage AX.25, quite on the contrary. I am trying to help it be put to rest without causing too much disruption, because at this point, it is very much an outdated system.

**Reply by SpudGunMan** [reply]
Ok, thank you. out dated as it might be it still functions wonderfully in the physical to include extra terrestrial telemetry. I was on a hunt for a BPQ alternative but think it’s best suited for radio use. This project might be better for other AX25 stack projects I’m sure but I was looking for winlink and aprs. 

I checked out that Reticulum project looks fancy but also can’t figure out how it applies to commonly used ham radio applications. Don’t think the FCC will be allowing use of ipv6 tomorrow. Thank you very much for your time and 73!

**Reply by markqvist** [reply]
You're welcome!

There's no issues from the FCC side with running IPv6 over amateur radio links, it's just not very efficient to do so ;) IPv6 is not designed for such low-bandwidth mediums, especially considering how "chatty" it is even when completely idle.

Reticulum is not build on IPv6, it is an entirely new stack. It can run over any sort of raw medium, packet radio included, and has no dependencies on IP or even Ethernet. It was designed to allow full networking functionality even over very slow mediums, such as packet radio.

One of the big advantages is that you can easily run it on top of IP if you want to though. So you can make hybrid systems that have links spanning both packet radio, private IP links, TCP over the Internet, LoRa radios, serial lines and more, all automatically connected into one coherent network.

And totally get that, sending AX.25 frames from microcontrollers and such is neat and simple. That is a huge plus. But it is still not a very reliable solution, especially if you care about the *integrity* of that telemetry.

**Reply by SpudGunMan** [reply]
for sure just used the IPv6 as an example of complications in FCC/rules which also appear quickly in published projects and all, I will check out the Reticulum .. smells interesting for including a shim to AGW emulator, dire wolf code might snip down easy ...

side note1; I included the project in my newly released beta "AUR-esk" ham app bapi. to get more visibility and testing as I haven't seen much use of kiss2tcp yet in the wild only white papers, with some shims to the past it could make a nice ARRL article to spring forward.

side note2; you see the white paper on vara full speed tcp/ip? @ vara'groups.io

**Reply by markqvist** [reply]
Ah, get it. Yeah, not going to contest that complications are usually a part of reality in regards to the FCC ;) Yes, just go ahead and play around with it however you like of course. Interesting stuff! Could you link me the app, would like to see what you are up to.

No, I should actually read the whitepaper. Have only looked at the stuff superficially because I have been busy with other things, but vara is very impressive!

**Reply by SpudGunMan** [reply]
[bapi](https://github.com/SpudGunMan/bapi) the current beta has "core apps" 
- tncattach is still in discovery phase now, haha! I will add it to core soon. 

if I trip over code to shim anything cool I will make sure to post back here!

vara thing, yea its cool the project appears stalled however or discontinued due to some very interesting decisions of soundmodem   use etc.. code out there to sniff around anyway. :)

**Reply by markqvist** [reply]
Nice project, good work. Thanks for making and sharing it!

**Reply by SpudGunMan** [reply]
Thanks :) 

Just saw this project and it may be of interest as well. https://radiomail.app/

**Comment by markqvist** [comment, marked as answer]
I am gong to collect a few different comments together here, and mark it as an aswer for future references on this topic:

The tnattach program lets you attach TNC devices as generic Ethernet/IP interfaces. It is not for using with "legacy" AX.25 apps, but for using KISS TNCs as Ethernet adapters / IP bridges.

Legacy packet apps don't use TCP in any way, they use AX.25. This is the case for all apps like Xastir, Winlink, etc.

Practically all apps allow you to connect directly to a TNC using KISS framing over a serial port or similar. Some apps allow you to connect to the TNC itself over TCP, ie., the TNC uses a TCP connection to expose it's KISS port, instead of a USB/serial connection.

Most legacy packet radio apps can also use AX.25 support built into the Linux kernel. In that case, the user-space programs talk to the kernel interface, usually called something like ax0 or ax1, and the kernel coordinates with the actual hardware TNC using KISS framing over either USB, serial or TCP.

Windows did not have actual kernel support for AX.25, so a clever op, SV2AGW, made the AGW program as a hack for Windows to allow multiple programs to share the same physical/virtual TNC. Linux never needed this, since there was direct support for AX.25 and device/app multiplexing in the kernel.

But now most distros are dropping the AX.25 layers from the kernel because the implementations are old, quite messy at this point, and more or less unmaintained.

This means, that while most legacy packet radio apps will still work when directly talking to a TNC, it is no longer possible to set up a multiplexed environment with many apps talking to the same TNC, like with the old ax25-utils included in Linux, until recently.

Reticulum of course let's you do this, and in a way much more flexible than the AX.25 stack ever did.

The tncattach program was never meant to remedy this situation. In fact, it was more meant to help transition away from AX.25.

While I have great respect for the pioneering work of AX.25, and all the utility and usefulness it has provided, I would also suggest moving away from it entirely by now.

In AX.25, there is no actual routing, it is very inefficient with bandwidth, it is completely unreliable, since no packets can be verified or authenticated, it is completely open to modification of traffic, eavesdropping and tampering, you must have complete and full trust in all nodes on the network, and many more problems that are simply a deal-breaker for any real-world use, even for hobbyists.

I think a much better solution is to use Reticulum, and help transition applications and use-cases to a much more modern, flexible and efficient packet-oriented stack instead.

The tncattach program allows you to run full Ethernet over packet radio, or simpler IP-only P2P links if that is more suitable to the situation. That is the intention, and what it delivers.

I am not trying to salvage AX.25, quite on the contrary. I am trying to help it be put to rest without causing too much disruption, because at this point, it is very much an outdated system.

**Reply by SpudGunMan** [reply]
Might be good to get published in ARRL etc to get attention, and get the fcc rules ironed out etc.

**Reply by markqvist** [reply]
In regards to using `tncattach` with radio systems over Amateur Radio spectrum, there really is no issues. It complies fully with FCC Part 97, as long as you remember to set the correct beacon/identification settings (see the help section for how to do that). Running Ethernet or IP (v4 or v6) on top of that is also not an issue in itself. You *could* run into issues with FCC regulations depending what kind of traffic you pass over those IP links, but that is a whole other disucssion.

Regarding FCC regulations and _Reticulum_ itself (which uses neither IP nor `tncattach`), I wrote in a bit more detail specifically in this post: https://github.com/markqvist/Reticulum/discussions/70#discussioncomment-3029969

**Reply by markqvist** [reply]
And just to make it clear to anyone reading this in the future, and avoid any possible confusion, the `tncattach` program is not related to, or used by Reticulum.

As mentioned above, it is a program written to replace the missing capability of using KISS-compatible packet radio modems as generic IP and Ethernet devices on Linux computers, that disappeared when distros started dropping support for the `ax25-utils` packages, and AX.25 kernel support.
